### PR TITLE
stage2 ARM: optimize airSliceElemVal for elem_size 1 or 4

### PR DESCRIPTION
**Pull request by joachimschmidt557**
In these cases, the AIR inst can be lowered to only one ldr instruction.

Also fixes shifts in arm.bits.Offset

### CODE REVIEW

File: src/arch/arm/CodeGen.zig [hunk 1]
```diff
@@ -1246,61 +1249,67 @@ fn airSliceElemVal(self: *Self, inst: Air.Inst.Index) !void {
...
-            self.register_manager.unfreezeRegs(&.{ base_mcv.register, offset_mcv.register });
+                try self.genArmBinOpCode(addr_reg, base_mcv, offset_mcv, false, .add, .unsigned);
 
-            try self.load(dst_mcv, .{ .register = addr_reg }, slice_ptr_field_type);
+                // I know we will unfreeze these registers at the end of
+                // the scope of :result. However, at this point in time,
+                // neither the base register nor the offset register
+                // contains any valuable data anymore. In order to reduce
+                // register pressure, unfreeze them prematurely
+                self.register_manager.unfreezeRegs(&.{ base_mcv.register, offset_mcv.register });
```
Comment (kubkon) [inline src/arch/arm/CodeGen.zig:1307]: Hmm, this doesn't feel right to me. We should perhaps figure out a way to do this automatically. Also, we only make a call to `load` subsequently, so I guess reducing pressure here won't have a measurable effect?

Comment (joachimschmidt557) [reply src/arch/arm/CodeGen.zig:1307]: Yes, this does in fact have a measurable impact: Without this, the register manager will quickly run out of usable registers (of which there are only 6, see code snippet below) in certain test cases which involve inline `memcpy`.

https://github.com/ziglang/zig/blob/95fbce2b958395a367a82ce33170edd93e686173/src/arch/arm/bits.zig#L189

The long term solution to this consists of

- calling extern `memcpy`
- use caller-preserved registers as well

But for now (and possibly also for the future for reducing register pressure), this tiny optimization really does help.

Comment (kubkon) [reply src/arch/arm/CodeGen.zig:1307]: Right, agreed. So in this case, I believe your intent will be better communicated if you remove the defers in favour of explicit management of unfreezing (in this block). Then, there is no need for the comment here. Rather, we could add a comment to this effect to the `RegisterManager`'s docs to better document different idioms and uses. Lemme know what you think about this proposal!

Comment (joachimschmidt557) [reply src/arch/arm/CodeGen.zig:1307]: Ok, will do!

Comment (kubkon) [reply src/arch/arm/CodeGen.zig:1307]: @joachimschmidt557 I'm gonna merge this as-is and you can follow up with our proposal in a subsequent PR. Hope that sounds good to you!

**Pull request merged**